### PR TITLE
Stat dist slepc

### PR DESCRIPTION
**Pull request by michalk8**
Hi @msmdev ,

@Marius1311 encounted a matrix where the ARPACK solver in `scipy.sparse.eigs` takes very long for rel. small matrix, see the linked issue below. I've included a SLEPc version, which will be tried first (if PETSc/SLEPc is present), and fall back to 
scipy if not. Still have to write tests for this.

closes #https://github.com/theislab/cellrank/issues/518

### CODE REVIEW

**Comment by msmdev** [comment]
Hi @michalk8, some test are failing, so I guess this is work in progress? :)

File: pygpcca/utils/_utils.py [hunk 1]
```diff
@@ -220,10 +220,54 @@ def _sdd(P: np.ndarray) -> np.ndarray:
...
+
+
 @stationary_distribution.register(spmatrix)
 def _sds(P: spmatrix) -> np.ndarray:
     # get the top two eigenvalues and vecs so we can check for irreducibility
-    vals, vecs = eigs(P.transpose(), k=2, which="LR", ncv=None)
+    try:
+        vals, vecs = _eigs_slepc(P.T, k=2, which="LR")
+    except ImportError:
+        vals, vecs = eigs(P.T, k=2, which="LR", ncv=None)
```
Comment (Marius1311) [inline pygpcca/utils/_utils.py:270]: I suggest a modified way of testing for irreducibility @michalk8. Let's use the test we use in Cellrank, i.e. using networkx to check whether there's more than one connected component. That's the standard way of testing for irreducibility and runs in linear time. The test we have here is something I came up with myself and is a bit more shaky - for example, it depends on these thresholds we have to set below. 

If we use networkx, we don't have to compute the top two eigenvalues but only the top (with largest real part) and we can also get rid of the sorting below (`# sort by real part and take the top one`)

**Review by Marius1311 — CHANGES_REQUESTED** [review]
I would change the way in which we check for irreducibility, see my comment.

File: pygpcca/utils/_utils.py (hunk 1 above)
Comment (Marius1311) [reply pygpcca/utils/_utils.py:270]: Okay never mind, @michalk8 convinced me that this is not necessary - we can leave this check as it is.

Comment (Marius1311) [reply pygpcca/utils/_utils.py:270]: One thing I would add though is adding some regression tests for this. I have some example matrices we can use here.

Comment (michalk8) [reply pygpcca/utils/_utils.py:270]: I'd say not to include any more depenencies if not necessary. On the matrix you've provided, it takes ~3 to get the stat. dist.

**Comment by michalk8** [comment]
> Hi @michalk8, some test are failing, so I guess this is work in progress? :)

@msmdev should be fixed now, ready for review.

**Comment by Marius1311** [comment]
Okay, so the current way in which we compute the stationary distribution is suboptimal I think. In an ideal world, we would use the schur decomposition we have already computed and obtain the top eigenvalues and eigenvectors from there - this would require no extra computational effort. 

Currently, depending on whether the matrix is sparse or dense, we do the following:
1. Matrix is sparse
  2. Try using SLEPSc to compute the top 2 eigenvalues and eigenvectors [CHECK UNIQUENESS]
  3. If no SLEPSC installed, use `scipy.linalg.eigs` to compute the top 2 eigenvalues and eigenvectors [CHECK UNIQUENESS]

**Review by Marius1311 — APPROVED** [review]
This fixes the original issue.

**Comment by Marius1311** [comment]
Okay, finally done, tests passing, ready for your review @msmdev.

**Comment by Marius1311** [comment]
Hi @msmdev, did you have a chance to check this out yet? Our new CellRank 1.3 release depends on this, so it would be great to merge this soon.

**Comment by msmdev** [comment]
> Hi @msmdev, did you have a chance to check this out yet? Our new CellRank 1.3 release depends on this, so it would be great to merge this soon. Thanks!

Hi @Marius1311, sorry for the delay! I will do it within this week - if it suffices?

**Comment by Marius1311** [comment]
Hi @msmdev, we're planning to do our release later today - can you please have a quick look over this today? Tests are passing and this is not a big change, it only makes sure we use SLEPSc to compute the stationary distribution when it's installed, that's it.

**Comment by msmdev** [comment]
@Marius1311, I fear today isn't very likely, since there are several urgent high priority tasks on my list (I was sick last week...) :(
It is not just taking a glance, but also doing a new release, before you can make real use of the changes, right?

**Comment by Marius1311** [comment]
Okay, no worries. We can also split the work - you just do the review and I do the new release (I guess I should have the permissions). But no rush, we can have the new CellRank 1.3 release without this today and then include this PR in a patch release to CellRank later this week.

**Comment by Marius1311** [comment]
Hi @msmdev, what's the status here?

**Comment by msmdev** [comment]
> @msmdev this PR fixes the original issue: for a sparse matrix, we now use by default SLEPSc to compute the stationary distribution, which has practical advantages in terms of scalability and numerical stability. This also adds some tests which all pass, so I would recommend to merge this.
> 
> For the future, I think we should however rethink the way in which we do this because the current way in which we compute the stationary distribution is suboptimal in my opinion. In an ideal world, we would use the schur decomposition we have already computed and obtain the top eigenvalues and eigenvectors from there - this would require no extra computational effort.

Unfortunately, the Schur vectors span the same subspace as the eigenvectors, but they aren't the eigenvectors.

> Currently, depending on whether the matrix is sparse or dense, we do the following:
> 
> * Matrix is sparse
>   
>   * Try using SLEPSc to compute the top 2 eigenvalues and eigenvectors [CHECK UNIQUENESS]
>   * If no SLEPSC installed, use `scipy.sparse.linalg.eigs` to compute the top 2 eigenvalues and eigenvectors [CHECK UNIQUENESS]
> * Matrix is dense
>   
>   * Try using the backward iteration algorithm.
>   * If that did not converge, use `scipy.linalg.eig` to compute the top eigenvector
> 
> As indicated, currently, we only check for uniqueness of the stationary distribution if the matrix is sparse. I suggest we do one of the following two options:

I wasn't aware of that: uniqueness should be checked in any case - at least, if the Markov chain is non-ergodic. How do you check for ergodicity (rreducibility and aperiodicity) in cellrank - mb. we could use the same way?
 
> 1. Get rid of the `backwars_iteration` algorithm, it took ages in my tests anyways. Always compute stationary distribution via spectral decomposition, using SLEPSc/`scipy.sparse.linalg.eigs` if the matrix is sparse and `scipy.linalg.eig` if it's dense, always compute (at least) the top two eigenvalues and use these to check for uniqueness of the stationary distribution

What was the reason for using backwards_interation anyway?

> 2. Include this in the routine that computes the schur decomposition, use the schur decomposition to get the required eigenvalues and vectors.

See my comment above...

**Comment by msmdev** [comment]
@Marius1311, could you please write a release note in RTD like in cellrank? I guess we should already have started doing this before, but I forgot it in the hassle. Also the only patch so far was very small - so no harm.

After you have done this and answered on my questions above, I can merge this and do the release today :)

**Comment by Marius1311** [comment]
> Unfortunately, the Schur vectors span the same subspace as the eigenvectors, but they aren't the eigenvectors.
> But I guess you mean using SLEPc to compute the eigenvectors?

No, actually I was mistaken here. I do not mean using SLEPc to compute the eigenvectors *in the future* - that's taken care of in this PR already. I actually meant using the Schur decomposition to get the eigendecomposition because I thought this would be easily possible - it turns out I confused this with using the real Schur decomposition to get the eigenvalues, which is easily possible by converting it to the complex Schur decomposition and reading off the diagonal. So, I was wrong about this one. 


> I wasn't aware of that: uniqueness should be checked in any case - at least, if the Markov chain is non-ergodic.
> How exactly do you check for uniqueness currently here (you said something in a comment that the current check is somewhat "shaky")?

Actually, the current check is much better than I thought originally and works very robustly in practice. I compute the top two eigenvalues and if both of them are very close to 1, than the matrix won't have a unique stationary distribution. 


> How do you check for ergodicity (irreducibility and aperiodicity) in cellrank - mb. we could use the same way?
> 

We use `networkx` to compute the strongly connected components. While this is O(n) in the number of states, @michalk8 voted against introducing an extra package dependency, so we decided not to include this here. 

> What was the reason for using backwards_iteration anyway?

That is a left over from msmtools. I have no idea. There are probably some situations in which it's faster if you don't have SLEPc installed. SLEPSc will **always** beat it.

**Comment by msmdev** [comment]
> > Unfortunately, the Schur vectors span the same subspace as the eigenvectors, but they aren't the eigenvectors.
> > But I guess you mean using SLEPc to compute the eigenvectors?
> 
> No, actually I was mistaken here. I do not mean using SLEPc to compute the eigenvectors _in the future_ - that's taken care of in this PR already. I actually meant using the Schur decomposition to get the eigendecomposition because I thought this would be easily possible - it turns out I confused this with using the real Schur decomposition to get the eigenvalues, which is easily possible by converting it to the complex Schur decomposition and reading off the diagonal. So, I was wrong about this one.
> 
> > I wasn't aware of that: uniqueness should be checked in any case - at least, if the Markov chain is non-ergodic.
> > How exactly do you check for uniqueness currently here (you said something in a comment that the current check is somewhat "shaky")?
> 
> Actually, the current check is much better than I thought originally and works very robustly in practice. I compute the top two eigenvalues and if both of them are very close to 1, than the matrix won't have a unique stationary distribution.

I see, I would also say that this is a sound way of checking for reducibility: if there are two or more eigenvalues (nearly) equal 1, the graph apart in two or more disconnected regions - so there is no unique stat. dist. 
I'm just not entirely sure about aperiodicity here - to have a unique stationary distribution we need irreducibility and aperiodicity. What are your thoughts here?

> > How do you check for ergodicity (irreducibility and aperiodicity) in cellrank - mb. we could use the same way?
> 
> We use `networkx` to compute the strongly connected components. While this is O(n) in the number of states, @michalk8 voted against introducing an extra package dependency, so we decided not to include this here.

I also would prefer not to include more external packages, if possible.
 
> > What was the reason for using backwards_iteration anyway?
> 
> That is a left over from msmtools. I have no idea. There are probably some situations in which it's faster if you don't have SLEPc installed. SLEPSc will **always** beat it.

**Comment by Marius1311** [comment]
Two comments on this, all set in the context of discrete and finite state-space MCs (our case): 
- actually, if there are `n` 1-eigenvalues, that does not mean the graph has to be disconnected. It just mean there are `n` strongly connected components, or, in the language of Markov chains, `n` recurrent classes, i.e. `n` regions you won't leave again, once you enter them. 
- Aperiodicity is not required to have a unique stationary distribution. It follows directly from Perron-Frobenius that for a row-stochastic matrix, irreducibility is enough to get a unique stationary distribution. However, if the matrix is also aperiodic, the stationary distribution corresponds to the **limiting distribution**. I think we have discussed this before, however, I'm happy to have a zoom-chat over this. I find this is well explained in e.g. Tolver 2016, "An introduction to Markov chains" or also Murphy 2012, "Machine learning: a probabilistic perspective". Let `A` be a row-stochastic, irreducible matrix with period `p`. `A` will then have `p` eigenvalues of magnitude 1 which will all be roots of unity, but there will be only 1 unique eigenvalue of real part 1 and zero imaginary part. That's called the Perron-Frobenius eigenvalue of `A` and it's simple, i.e. has multiplicity 1. Therefore, if we define the stationary distribution via `\pi^T A = \pi^T`, then the solution to this equation is unique, even if `A` has period `p`.

**Comment by msmdev** [comment]
> Two comments on this, all set in the context of discrete and finite state-space MCs (our case):
> 
> * actually, if there are `n` 1-eigenvalues, that does not mean the graph has to be disconnected. It just mean there are `n` strongly connected components, or, in the language of Markov chains, `n` recurrent classes, i.e. `n` regions you won't leave again, once you enter them.

Sorry, you are right - my wording was imprecise on this. Also to be really precise: 1-eigenvalues means real ones here...

> * Aperiodicity is not required to have a unique stationary distribution. It follows directly from Perron-Frobenius that for a row-stochastic matrix, irreducibility is enough to get a unique stationary distribution. However, if the matrix is also aperiodic, the stationary distribution corresponds to the **limiting distribution**. I think we have discussed this before, however, I'm happy to have a zoom-chat over this. I find this is well explained in e.g. Tolver 2016, "An introduction to Markov chains" or also Murphy 2012, "Machine learning: a probabilistic perspective". Let `A` be a row-stochastic, irreducible matrix with period `p`. `A` will then have `p` eigenvalues of magnitude 1 which will all be roots of unity, but there will be only 1 unique eigenvalue of real part 1 and zero imaginary part. That's called the Perron-Frobenius eigenvalue of `A` and it's simple, i.e. has multiplicity 1. Therefore, if we define the stationary distribution via `\pi^T A = \pi^T`, then the solution to this equation is unique, even if `A` has period `p`.

Guess you are right here as well, rethinking this and trying to remember my own writing in my PhD-Thesis. XD
I confused uniqueness of the stat. dist. `pi` with ergodicity here: >> MC ergodic -> unique << `pi`, but **not** the opposite. >> `pi` unique -> MC ergodic << isn't true.

**Comment by msmdev** [comment]
@Marius1311, just one thought regarding real 1 eigenvalues: to be associated with with strongly connected components the 1's need to be real and positive, not only real. If they are real and negative, they will be associated with real positive 1's they are paired with (second roots of unity) building 2-cycles. So looking for real 1's with no imaginary part doesn't suffice to exclude cycles, one has to check for sign as well.

**Comment by msmdev** [comment]
> @Marius1311, just one thought regarding real 1 eigenvalues: to be associated with with strongly connected components the 1's need to be real and positive, not only real. If they are real and negative, they will be associated with real positive 1's they are paired with (second roots of unity) building 2-cycles. So looking for real 1's with no imaginary part doesn't suffice to exclude cycles, one has to check for sign as well.

This is just a side note - not touching your test, which is safe, since you are using `method='LR'`.

File: docs/source/release_notes.rst [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. rubric:: Bugfixes
+
+- Fix not catching ``ArpackError`` when computing stationary distribution and mypy-related linting issues
+  `PR 21 <https://github.com/msmdev/pyGPCCA/pull/21>`_.
+
+.. rubric:: Improvements
+
+- Use PETSc/SLEPc, if installed, to compute stationary distribution speed it up
```
Comment (msmdev) [inline docs/source/release_notes.rst:19]: Please correct grammar, e.g. into "... to speed up the computation of the stationary distribution"

**Review by msmdev — CHANGES_REQUESTED** [review]

File: docs/source/release_notes.rst (hunk 2 above)
Comment (michalk8) [reply docs/source/release_notes.rst]: Thanks for spotting this, itis fixed now.

**Review by msmdev — APPROVED** [review]

**Comment by msmdev** [comment]
@Marius1311, could you please open an issue regarding missing irreducibility checks for dense transition matrices please?

**Comment by Marius1311** [comment]
yes, will do! Thanks for your help @michalk8 and thanks for looking into this @msmdev

**Comment by msmdev** [comment]
@michalk8 , @Marius1311 , the release is deployed in pypi now, but not on conda-forge... 
@michalk8 could you mb. take care of this (since it seems to have more downloads than pypi)?

**Comment by msmdev** [comment]
@michalk8, mb. also explain the process to me (again? don't remember... %) - so can (hopefully) take care of it next time XD

**Comment by msmdev** [comment]
Strange: CI #149 passes and is deployed but #150 fails.

**Comment by michalk8** [comment]
@msmdev 

It should be done automatically, see [here](https://conda-forge.org/docs/maintainer/updating_pkgs.html#pushing-to-regro-cf-autotick-bot-branch)
Some intervention from maintainer side (yours) need to happen only when e.g. package version changes.

**Comment by msmdev** [comment]
> Strange: CI #149 passes and is deployed but #150 fails.

Ok, rerun the tests - now it passes.

**Comment by msmdev** [comment]
> @msmdev
> 
> It should be done automatically, see [here](https://conda-forge.org/docs/maintainer/updating_pkgs.html#pushing-to-regro-cf-autotick-bot-branch)
> Some intervention from maintainer side (yours) need to happen only when e.g. package version changes.

Ok, might just take some time.

**Comment by msmdev** [comment]
@michalk8, @Marius1311, I will be on Easter holiday next week btw. - so I won't be available. Have a nice Easter, too!

**Comment by Marius1311** [comment]
thanks!

**Comment by michalk8** [comment]
> @michalk8, @Marius1311, I will be on Easter holiday next week btw. - so I won't be available. Have a nice Easter, too!

Thanks, likewise!

**Pull request merged**